Blsct arith classes with tests

gogoex opened the pull request:
For issue #42 

- Reviewed the original implementation of Scalar and G1Point and made necessary changes
- Added tests for each exposed methods of Scalar and G1Point
- Added Elements class that represents a vector of Scalar and G1Point classes
- Added operator overloads to Scalar, G1Point and Elements classes to improve readability of code
- Confirmed that Scalar, G1Point and Elements provide enough functionalities to implement a range proof by adding a test that implemented a range proof

NeblioTeam commented on src/blsct/arith/elements.cpp:
```diff
@@ -0,0 +1,232 @@
+        return ret;
+
+    } else {
+        throw std::runtime_error("Now implemented");
```
Small typo, I think it should be `Not` instead of `Now`

NeblioTeam commented on src/blsct/arith/elements.cpp:
```diff
@@ -0,0 +1,232 @@
+
+    bool ret = true;
+    for (size_t i = 0; i < m_vec.size(); ++i) {
+        ret = ret && (m_vec[i] == other[i]);
```
Suggestion: It could be more efficient to do 

```c++
if (m_vec[i] != other[i]) return false;
```

aguycalled commented on src/blsct/arith/elements.cpp:
```diff
@@ -0,0 +1,232 @@
+        }
+        return ret;
+    } else {
+        throw std::runtime_error("Now implemented");
```
I think it should be `Not` instead of `Now`.

gogoex replied:
fixed the typo!

aguycalled commented on src/blsct/arith/elements.cpp:
```diff
@@ -0,0 +1,232 @@
+        return ret;
+
+    } else {
+        throw std::runtime_error("Now implemented");
```
I think it should be `Not` instead of `Now`.

gogoex replied:
fixed the typo!

aguycalled commented on src/blsct/arith/elements.cpp:
```diff
@@ -0,0 +1,232 @@
+        return ret;
+
+    } else {
+        throw std::runtime_error("Now implemented");
```
I think it should be `Not` instead of `Now`.

gogoex replied:
fixed the typo again!

aguycalled commented on src/blsct/arith/elements.cpp:
```diff
@@ -0,0 +1,232 @@
+
+    bool ret = true;
+    for (size_t i = 0; i < m_vec.size(); ++i) {
+        ret = ret && (m_vec[i] == other[i]);
```
I think it would be more efficient to do

```c++
if (m_vec[i] != other[i]) return false;
```

gogoex replied:
right. fixed.

aguycalled commented on src/blsct/arith/g1point.cpp:
```diff
@@ -0,0 +1,217 @@
+    return temp;
+}
+
+G1Point G1Point::MulVec(const std::vector<G1Point>& g_vec, const std::vector<Scalar>& s_vec)
```
Could `Elements<G1Point>` and `Elements<Scalar>` be used here, for consistency?

gogoex replied:
i tried, but it seems like this it creates a circular dependency and compilation fails. perhaps we can let Elements absorb MulVec functions and let it use mclBnG1_mulVec, and drop MulVec functions G1Point? let me try so that you can take a look at it.

gogoex replied:
i migrated MulVec that takes vectors of G1Points and Scalas to Elements. can you check? i couldn't migrate MulVec that takes mcl types since it doesn't use the template parameter, so i left it in G1Point

gogoex replied:
and the migrated MulVec now looks like

```C++
G1Point MulVec(const Elements<Scalar>& scalars) const;
```

aguycalled replied:
looking good to me

aguycalled commented on src/blsct/arith/g1point.cpp:
```diff
@@ -0,0 +1,217 @@
+
+G1Point G1Point::MulVec(const std::vector<G1Point>& g_vec, const std::vector<Scalar>& s_vec)
+{
+    if (g_vec.size() != s_vec.size()) {
```
If the previous comment applies, `ConfirmSizesMatch` could be used here, I think.

gogoex replied:
migrated MulVec now uses it

aguycalled commented:
Getting an error when building locally:

```
blsct/arith/elements.cpp:115:17: error: 'm_vec' is a private member of 'Elements<G1Point>'
            ret.m_vec.push_back(m_vec[i] * other[i]);
                ^
blsct/arith/elements.cpp:123:45: note: in instantiation of member function 'Elements<Scalar>::operator*' requested here
template Elements<Scalar> Elements<Scalar>::operator*(const Elements<Scalar>& other) const;
                                            ^
./blsct/arith/elements.h:84:20: note: declared private here
    std::vector<T> m_vec;
                   ^
blsct/arith/elements.cpp:108:17: error: 'm_vec' is a private member of 'Elements<Scalar>'
            ret.m_vec.push_back(m_vec[i] * other[i]);
                ^
blsct/arith/elements.cpp:124:47: note: in instantiation of member function 'Elements<G1Point>::operator*' requested here
template Elements<G1Point> Elements<G1Point>::operator*(const Elements<Scalar>& other) const;
                                              ^
./blsct/arith/elements.h:84:20: note: declared private here
    std::vector<T> m_vec;
                   ^
blsct/arith/elements.cpp:139:17: error: 'm_vec' is a private member of 'Elements<G1Point>'
            ret.m_vec.push_back(m_vec[i] * s);
                ^
blsct/arith/elements.cpp:147:45: note: in instantiation of member function 'Elements<Scalar>::operator*' requested here
template Elements<Scalar> Elements<Scalar>::operator*(const Scalar&) const;
                                            ^
./blsct/arith/elements.h:84:20: note: declared private here
    std::vector<T> m_vec;
                   ^
blsct/arith/elements.cpp:132:17: error: 'm_vec' is a private member of 'Elements<Scalar>'
            ret.m_vec.push_back(m_vec[i] * s);
                ^
blsct/arith/elements.cpp:148:47: note: in instantiation of member function 'Elements<G1Point>::operator*' requested here
template Elements<G1Point> Elements<G1Point>::operator*(const Scalar& s) const;
                                              ^
./blsct/arith/elements.h:84:20: note: declared private here
    std::vector<T> m_vec;
                   ^
21 warnings and 4 errors generated.
make[2]: *** [blsct/arith/libbitcoin_node_a-elements.o] Error 1
```

Environment:

```
Apple clang version 13.1.6 (clang-1316.0.21.2)
Target: arm64-apple-darwin21.2.0
```

gogoex commented:
it seems that gcc allows accessing private member of the same template class regardless of its instantiation type is, but clang is more strict. i made the member vetor variable public since G1Point ans Scalar do so as well. i confirmed that now the code builds on clang environment

aguycalled commented:
> it seems that gcc allows accessing private member of the same template class regardless of its instantiation type is, but clang is more strict. i made the member vetor variable public since G1Point ans Scalar do so as well. i confirmed that now the code builds on clang environment

confirmed it now builds locally on my MacBook Pro 👍

aguycalled commented:
One of the tests fails for me:

```
% ./src/test/test_navcoin 
Running 587 test cases...
test/blsct/arith/integration_tests.cpp:30: error: in "test_integration_invert_zero": check x.Pow(zero) == x.Pow(zero.Invert()) has failed

*** 1 failure is detected in the test module "Navcoin Core Test Suite"
```

gogoex commented:
@aguycalled The test was originally created to test FirstNPow for a case where y is 0:
```c++
    // (64)
    auto hhp = hh * Scalars::FirstNPow(n, y.Invert());
```
but now i think the test is not necessary since y is in Z^*_p and non-zero.

Also since multiplying any element by its invese results in unit element 1 and multiplying any element by 0 results in 0, I think the inverse of 0 in Z_p is undefined. It probably just happes to return 0 on x86_64 implementation.

So I removed the test and added another test case for Scalars::FirstNPow to test if it always return alist that starts with 1 regardless of the value of base value instead. 

Can you try again?

aguycalled commented:
> @aguycalled The test was originally created to test FirstNPow for a case where y is 0:
> 
> ```c++
>     // (64)
>     auto hhp = hh * Scalars::FirstNPow(n, y.Invert());
> ```
> 
> but now i think the test is not necessary since y is in Z^*_p and non-zero.
> 
> Also since multiplying any element by its invese results in unit element 1 and multiplying any element by 0 results in 0, I think the inverse of 0 in Z_p is undefined. It probably just happes to return 0 on x86_64 implementation.
> 
> So I removed the test and added another test case for Scalars::FirstNPow to test if it always return alist that starts with 1 regardless of the value of base value instead.
> 
> Can you try again?

All the tests pass for me now 👍 

```zsh
% ./src/test/test_navcoin                       
Running 586 test cases...

*** No errors detected
```

Maybe, `Scalar::Invert()` should throw an exception if the value is `zero`.

mxaddict commented:
> 

I think this would be a good idea, since we are not supposed to use a zero value in Scalar right?

gogoex commented:
sure. i will make the input 0 throw an exception

aguycalled reviewed (approved):
Approving and merging. Pending solving the CI issues in a different task #61.
